patches/api/0387-Add-damage-source-wrapper.patch

```diff
@@ -0,0 +1,1070 @@
+@@ -0,0 +1,39 @@
++package io.papermc.paper.entity.damage;
++
++import org.bukkit.Bukkit;
++
++public final class StaticDamageOrigin {
++    // TODO <!> keeps this list up to date with vanilla
++    // Vanilla
++    public static final DamageOrigin IN_FIRE = fromNMS("inFire");
++    public static final DamageOrigin LIGHTNING_BOLT = fromNMS("lightningBolt");
++    public static final DamageOrigin ON_FIRE = fromNMS("onFire");
++    public static final DamageOrigin LAVA = fromNMS("lava");
++    public static final DamageOrigin HOT_FLOOR = fromNMS("hotFloor");
++    public static final DamageOrigin IN_WALL = fromNMS("inWall");
++    public static final DamageOrigin CRAMMING = fromNMS("cramming");
++    public static final DamageOrigin DROWN = fromNMS("drown");
++    public static final DamageOrigin STARVE = fromNMS("starve");
++    public static final DamageOrigin CACTUS = fromNMS("cactus");
++    public static final DamageOrigin FALL = fromNMS("fall");
++    public static final DamageOrigin FLY_INTO_WALL = fromNMS("flyIntoWall");
++    public static final DamageOrigin OUT_OF_WORLD = fromNMS("outOfWorld");
++    public static final DamageOrigin GENERIC = fromNMS("generic");
++    public static final DamageOrigin MAGIC = fromNMS("magic");
++    public static final DamageOrigin WITHER = fromNMS("wither");
++    public static final DamageOrigin ANVIL = fromNMS("anvil");
++    public static final DamageOrigin FALLING_BLOCK = fromNMS("fallingBlock");
++    public static final DamageOrigin DRAGON_BREATH = fromNMS("dragonBreath");
++    public static final DamageOrigin DRY_OUT = fromNMS("dryout");
++    public static final DamageOrigin SWEET_BERRY_BUSH = fromNMS("sweetBerryBush");
++    public static final DamageOrigin FREEZE = fromNMS("freeze");
++    public static final DamageOrigin FALLING_STALACTITE = fromNMS("fallingStalactite");
++    public static final DamageOrigin STALAGMITE = fromNMS("stalagmite");
++
++    // Cache this one
++    public static final DamageOrigin BAD_RESPAWN_POINT_EXPLOSION = fromNMS("badRespawnPoint");
++
++    private static DamageOrigin fromNMS(String key) {
++        return Bukkit.getUnsafe().getDamageOrigin(key);
++    }
++}
+diff --git a/src/main/java/io/papermc/paper/entity/damage/builder/AbstractDamageOriginBuilder.java b/src/main/java/io/papermc/paper/entity/damage/builder/AbstractDamageOriginBuilder.java
+new file mode 100644
+index 0000000000000000000000000000000000000000..bcbf108356ae834b67edf2f93718c7c5f5b4e5e5
+--- /dev/null
++++ b/src/main/java/io/papermc/paper/entity/damage/builder/AbstractDamageOriginBuilder.java
+@@ -0,0 +1,145 @@
++package io.papermc.paper.entity.damage.builder;
++
++import io.papermc.paper.entity.damage.DamageBypass;
++import net.kyori.adventure.text.Component;
++import org.bukkit.entity.LivingEntity;
++import org.jetbrains.annotations.NotNull;
++
++import java.util.function.Function;
++
++public abstract class AbstractDamageOriginBuilder<T, B> {
```

Comment:
This was discussed in my implementation but not actually implemented yet. 
Basically, you want to make these builders an interface and expose them via Bukkit.

This was the example I had.
```java
EntityDamageSource damageSource = Bukkit.newEntitySource("pricked", event.getPlayer())
            .setBypassInvul(true)
            .build();
```
Basically, you want it to wrap an NMS origin object (extend it)
Then, when calling build return the same wrapped NMS object except the interface that will represent it will be immutable. So, DamageBuilder builds an nms backed **DamageOrigin**.